// tinybench/dist/index.cjs
"use strict";
var v = Object.defineProperty;
var P = Object.getOwnPropertyDescriptor;
var C = Object.getOwnPropertyNames;
var H = Object.prototype.hasOwnProperty;
var j = (s, r, t) => r in s ? v(s, r, { enumerable: !0, configurable: !0, writable: !0, value: t }) : s[r] = t;
var z = (s, r) => {
  for (var t in r)
    v(s, t, { get: r[t], enumerable: !0 });
}, q = (s, r, t, e) => {
  if (r && typeof r == "object" || typeof r == "function")
    for (let n of C(r))
      !H.call(s, n) && n !== t && v(s, n, { get: () => r[n], enumerable: !(e = P(r, n)) || e.enumerable });
  return s;
};
var V = (s) => q(v({}, "__esModule", { value: !0 }), s);
var a = (s, r, t) => (j(s, typeof r != "symbol" ? r + "" : r, t), t);

// src/index.ts
var W = {};
z(W, {
  Bench: () => f,
  Task: () => p,
  default: () => U,
  hrtimeNow: () => M,
  now: () => w
});
module.exports = V(W);

// src/event.ts
function i(s, r = null) {
  let t = new Event(s);
  return r && Object.defineProperty(t, "task", {
    value: r,
    enumerable: !0,
    writable: !1,
    configurable: !1
  }), t;
}

// src/constants.ts
var $ = {
  1: 12.71,
  2: 4.303,
  3: 3.182,
  4: 2.776,
  5: 2.571,
  6: 2.447,
  7: 2.365,
  8: 2.306,
  9: 2.262,
  10: 2.228,
  11: 2.201,
  12: 2.179,
  13: 2.16,
  14: 2.145,
  15: 2.131,
  16: 2.12,
  17: 2.11,
  18: 2.101,
  19: 2.093,
  20: 2.086,
  21: 2.08,
  22: 2.074,
  23: 2.069,
  24: 2.064,
  25: 2.06,
  26: 2.056,
  27: 2.052,
  28: 2.048,
  29: 2.045,
  30: 2.042,
  31: 2.0399,
  32: 2.0378,
  33: 2.0357,
  34: 2.0336,
  35: 2.0315,
  36: 2.0294,
  37: 2.0273,
  38: 2.0252,
  39: 2.0231,
  40: 2.021,
  41: 2.0198,
  42: 2.0186,
  43: 2.0174,
  44: 2.0162,
  45: 2.015,
  46: 2.0138,
  47: 2.0126,
  48: 2.0114,
  49: 2.0102,
  50: 2.009,
  51: 2.0081,
  52: 2.0072,
  53: 2.0063,
  54: 2.0054,
  55: 2.0045,
  56: 2.0036,
  57: 2.0027,
  58: 2.0018,
  59: 2.0009,
  60: 2,
  61: 1.9995,
  62: 1.999,
  63: 1.9985,
  64: 1.998,
  65: 1.9975,
  66: 1.997,
  67: 1.9965,
  68: 1.996,
  69: 1.9955,
  70: 1.995,
  71: 1.9945,
  72: 1.994,
  73: 1.9935,
  74: 1.993,
  75: 1.9925,
  76: 1.992,
  77: 1.9915,
  78: 1.991,
  79: 1.9905,
  80: 1.99,
  81: 1.9897,
  82: 1.9894,
  83: 1.9891,
  84: 1.9888,
  85: 1.9885,
  86: 1.9882,
  87: 1.9879,
  88: 1.9876,
  89: 1.9873,
  90: 1.987,
  91: 1.9867,
  92: 1.9864,
  93: 1.9861,
  94: 1.9858,
  95: 1.9855,
  96: 1.9852,
  97: 1.9849,
  98: 1.9846,
  99: 1.9843,
  100: 1.984,
  101: 1.9838,
  102: 1.9836,
  103: 1.9834,
  104: 1.9832,
  105: 1.983,
  106: 1.9828,
  107: 1.9826,
  108: 1.9824,
  109: 1.9822,
  110: 1.982,
  111: 1.9818,
  112: 1.9816,
  113: 1.9814,
  114: 1.9812,
  115: 1.9819,
  116: 1.9808,
  117: 1.9806,
  118: 1.9804,
  119: 1.9802,
  120: 1.98,
  infinity: 1.96
}, b = $;

// src/utils.ts
var D = (s) => s / 1e6, M = () => D(Number(process.hrtime.bigint())), w = () => performance.now();
function G(s) {
  return s !== null && typeof s == "object" && typeof s.then == "function";
}
var O = (s, r) => s.reduce((e, n) => e + (n - r) ** 2, 0) / (s.length - 1) || 0, J = (async () => {
}).constructor, Q = (s) => s.constructor === J, F = async (s) => {
  if (Q(s.fn))
    return !0;
  try {
    if (s.opts.beforeEach != null)
      try {
        await s.opts.beforeEach.call(s);
      } catch (e) {
      }
    let r = s.fn(), t = G(r);
    if (t)
      try {
        await r;
      } catch (e) {
      }
    if (s.opts.afterEach != null)
      try {
        await s.opts.afterEach.call(s);
      } catch (e) {
      }
    return t;
  } catch (r) {
    return !1;
  }
};

// src/task.ts
var p = class extends EventTarget {
  constructor(t, e, n, o = {}) {
    super();
    a(this, "bench");
    a(this, "name");
    a(this, "fn");
    a(this, "runs", 0);
    a(this, "result");
    a(this, "opts");
    this.bench = t, this.name = e, this.fn = n, this.opts = o;
  }
  async loop(t, e) {
    var l;
    let n = 0, o = [];
    if (this.opts.beforeAll != null)
      try {
        await this.opts.beforeAll.call(this);
      } catch (h) {
        return { error: h };
      }
    let u = await F(this);
    try {
      for (; (n < t || o.length < e) && !((l = this.bench.signal) != null && l.aborted); ) {
        this.opts.beforeEach != null && await this.opts.beforeEach.call(this);
        let h = 0;
        if (u) {
          let c = this.bench.now();
          await this.fn.call(this), h = this.bench.now() - c;
        } else {
          let c = this.bench.now();
          this.fn.call(this), h = this.bench.now() - c;
        }
        o.push(h), n += h, this.opts.afterEach != null && await this.opts.afterEach.call(this);
      }
    } catch (h) {
      return { error: h };
    }
    if (this.opts.afterAll != null)
      try {
        await this.opts.afterAll.call(this);
      } catch (h) {
        return { error: h };
      }
    return { samples: o };
  }
  async run() {
    var n, o;
    if ((n = this.result) != null && n.error)
      return this;
    this.dispatchEvent(i("start", this)), await this.bench.setup(this, "run");
    let { samples: t, error: e } = await this.loop(this.bench.time, this.bench.iterations);
    if (this.bench.teardown(this, "run"), t) {
      let u = t.reduce((T, g) => T + g, 0);
      this.runs = t.length, t.sort((T, g) => T - g);
      let l = u / this.runs, h = 1e3 / l, c = t.length, m = c - 1, R = t[0], B = t[m], E = u / t.length || 0, k = O(t, E), y = Math.sqrt(k), x = y / Math.sqrt(c), A = b[String(Math.round(m) || 1)] || b.infinity, L = x * A, _ = L / E * 100, K = t[Math.ceil(c * 0.75) - 1], N = t[Math.ceil(c * 0.99) - 1], I = t[Math.ceil(c * 0.995) - 1], S = t[Math.ceil(c * 0.999) - 1];
      if ((o = this.bench.signal) != null && o.aborted)
        return this;
      this.setResult({
        totalTime: u,
        min: R,
        max: B,
        hz: h,
        period: l,
        samples: t,
        mean: E,
        variance: k,
        sd: y,
        sem: x,
        df: m,
        critical: A,
        moe: L,
        rme: _,
        p75: K,
        p99: N,
        p995: I,
        p999: S
      });
    }
    if (e) {
      if (this.setResult({ error: e }), this.bench.throws)
        throw e;
      this.dispatchEvent(i("error", this)), this.bench.dispatchEvent(i("error", this));
    }
    return this.dispatchEvent(i("cycle", this)), this.bench.dispatchEvent(i("cycle", this)), this.dispatchEvent(i("complete", this)), this;
  }
  async warmup() {
    var e;
    if ((e = this.result) != null && e.error)
      return;
    this.dispatchEvent(i("warmup", this)), await this.bench.setup(this, "warmup");
    let { error: t } = await this.loop(this.bench.warmupTime, this.bench.warmupIterations);
    if (this.bench.teardown(this, "warmup"), t && (this.setResult({ error: t }), this.bench.throws))
      throw t;
  }
  addEventListener(t, e, n) {
    super.addEventListener(t, e, n);
  }
  removeEventListener(t, e, n) {
    super.removeEventListener(t, e, n);
  }
  setResult(t) {
    this.result = { ...this.result, ...t }, Object.freeze(this.result);
  }
  reset() {
    this.dispatchEvent(i("reset", this)), this.runs = 0, this.result = void 0;
  }
};

// src/bench.ts
var f = class extends EventTarget {
  constructor(t = {}) {
    var e, n, o, u, l, h, c, m;
    super();
    a(this, "_tasks", /* @__PURE__ */ new Map());
    a(this, "_todos", /* @__PURE__ */ new Map());
    a(this, "signal");
    a(this, "throws");
    a(this, "warmupTime", 100);
    a(this, "warmupIterations", 5);
    a(this, "time", 500);
    a(this, "iterations", 10);
    a(this, "now", w);
    a(this, "setup");
    a(this, "teardown");
    this.now = (e = t.now) != null ? e : this.now, this.warmupTime = (n = t.warmupTime) != null ? n : this.warmupTime, this.warmupIterations = (o = t.warmupIterations) != null ? o : this.warmupIterations, this.time = (u = t.time) != null ? u : this.time, this.iterations = (l = t.iterations) != null ? l : this.iterations, this.signal = t.signal, this.throws = (h = t.throws) != null ? h : !1, this.setup = (c = t.setup) != null ? c : () => {
    }, this.teardown = (m = t.teardown) != null ? m : () => {
    }, this.signal && this.signal.addEventListener(
      "abort",
      () => {
        this.dispatchEvent(i("abort"));
      },
      { once: !0 }
    );
  }
  runTask(t) {
    var e;
    return (e = this.signal) != null && e.aborted ? t : t.run();
  }
  async run() {
    this.dispatchEvent(i("start"));
    let t = [];
    for (let e of [...this._tasks.values()])
      t.push(await this.runTask(e));
    return this.dispatchEvent(i("complete")), t;
  }
  async runConcurrently(t = 1 / 0) {
    this.dispatchEvent(i("start"));
    let e = [...this._tasks.values()], n = [];
    return await (async () => {
      for (; e.length > 0; ) {
        let u = [];
        for (; u.length < t && e.length > 0; ) {
          let h = e.pop();
          u.push(this.runTask(h));
        }
        let l = await Promise.all(u);
        n.push(...l);
      }
    })(), this.dispatchEvent(i("complete")), n;
  }
  async warmup() {
    this.dispatchEvent(i("warmup"));
    for (let [, t] of this._tasks)
      await t.warmup();
  }
  reset() {
    this.dispatchEvent(i("reset")), this._tasks.forEach((t) => {
      t.reset();
    });
  }
  add(t, e, n = {}) {
    let o = new p(this, t, e, n);
    return this._tasks.set(t, o), this.dispatchEvent(i("add", o)), this;
  }
  todo(t, e = () => {
  }, n = {}) {
    let o = new p(this, t, e, n);
    return this._todos.set(t, o), this.dispatchEvent(i("todo", o)), this;
  }
  remove(t) {
    let e = this.getTask(t);
    return e && (this.dispatchEvent(i("remove", e)), this._tasks.delete(t)), this;
  }
  addEventListener(t, e, n) {
    super.addEventListener(t, e, n);
  }
  removeEventListener(t, e, n) {
    super.removeEventListener(t, e, n);
  }
  table(t) {
    return this.tasks.map((e) => {
      if (e.result) {
        if (e.result.error)
          throw e.result.error;
        return (t == null ? void 0 : t(e)) || {
          "Task Name": e.name,
          "ops/sec": e.result.error ? "NaN" : parseInt(e.result.hz.toString(), 10).toLocaleString(),
          "Average Time (ns)": e.result.error ? "NaN" : e.result.mean * 1e3 * 1e3,
          Margin: e.result.error ? "NaN" : `\xB1${e.result.rme.toFixed(2)}%`,
          Samples: e.result.error ? "NaN" : e.result.samples.length
        };
      }
      return null;
    });
  }
  get results() {
    return [...this._tasks.values()].map((t) => t.result);
  }
  get tasks() {
    return [...this._tasks.values()];
  }
  get todos() {
    return [...this._todos.values()];
  }
  getTask(t) {
    return this._tasks.get(t);
  }
};

// src/index.ts
var U = f;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Bench,
  Task,
  hrtimeNow,
  now
});
